src/mame/msx/msx_keyboard.cpp

```diff
@@ -0,0 +1,828 @@
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_COMMA)     PORT_CHAR(',')           PORT_CHAR(';')
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_STOP)      PORT_CHAR('.')           PORT_CHAR(':')
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_SLASH)     PORT_CHAR('-')           PORT_CHAR('_')
+INPUT_PORTS_END
+
+
+static INPUT_PORTS_START(msx2_keypad)
+	PORT_MODIFY("KEY9")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_ASTERISK)  PORT_CHAR(UCHAR_MAMEKEY(ASTERISK))
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_PLUS_PAD)  PORT_CHAR(UCHAR_MAMEKEY(PLUS_PAD))
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_SLASH_PAD) PORT_CHAR(UCHAR_MAMEKEY(SLASH_PAD))
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_0_PAD)     PORT_CHAR(UCHAR_MAMEKEY(0_PAD))
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_1_PAD)     PORT_CHAR(UCHAR_MAMEKEY(1_PAD))
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_2_PAD)     PORT_CHAR(UCHAR_MAMEKEY(2_PAD))
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_3_PAD)     PORT_CHAR(UCHAR_MAMEKEY(3_PAD))
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_4_PAD)     PORT_CHAR(UCHAR_MAMEKEY(4_PAD))
+
+	PORT_MODIFY("KEY10")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_5_PAD)     PORT_CHAR(UCHAR_MAMEKEY(5_PAD))
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_6_PAD)     PORT_CHAR(UCHAR_MAMEKEY(6_PAD))
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_7_PAD)     PORT_CHAR(UCHAR_MAMEKEY(7_PAD))
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_8_PAD)     PORT_CHAR(UCHAR_MAMEKEY(8_PAD))
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_9_PAD)     PORT_CHAR(UCHAR_MAMEKEY(9_PAD))
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_MINUS_PAD) PORT_CHAR(UCHAR_MAMEKEY(MINUS_PAD))
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_ENTER_PAD) PORT_CHAR(UCHAR_MAMEKEY(COMMA_PAD))
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_DEL_PAD)   PORT_CHAR(UCHAR_MAMEKEY(DEL_PAD))
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2)
+	PORT_INCLUDE(msx)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2de)
+	PORT_INCLUDE(msxde)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2fr)
+	PORT_INCLUDE(msxfr)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2jp)
+	PORT_INCLUDE(msxjp)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2kr)
+	PORT_INCLUDE(msxkr)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2ru)
+	PORT_INCLUDE(msxru)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2sp)
+	PORT_INCLUDE(msxsp)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+
+INPUT_PORTS_START(msx2uk)
+	PORT_INCLUDE(msxuk)
+	PORT_INCLUDE(msx2_keypad)
+INPUT_PORTS_END
+
+}
```

Comment:
Please comment the closing brace of namespaces – since they’re often a very long way away from the opening brace, it’s useful when reading the file.